### DISCUSSION TITLE
Should the use of Body(.json) set the Content-Type accordingly automatically ?

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-url-routing#15] [Ideas]
Hi,

If we use a **Route** like this one. Should the `Content-Type` be set automatically to `application/json` ?

```swift
  Route(.case(BuildRoute.abort)) {
    Method.post
    Body(.json(BuildRoute.BodyParameters.self))
  }
```

It remove this `Headers { Field("Content-Type") { Parse { "application/json" } } }`

```swift
  Route(.case(BuildRoute.abort)) {
    Headers { Field("Content-Type") { Parse { "application/json" } } }
    Method.post
    Body(.json(BuildRoute.BodyParameters.self))
  }
```

Thanks,

**Comment by stephencelis** [comment]
@mackoj I think there's definitely something to this. And `FormData` should probably set a header, too.

I think we probably don't want to be so strict with the parsing, though. Instead, we could use `baseRequestData` to prefer printing this header.

```swift
Body(.json(BuildRoute.BodyParameters.self))
// =
Body(.json(BuildRoute.BodyParameters.self))
  .baseRequestData(.init(headers: ["Content-Type": ["application/json"]]))
```

**Reply by mackoj** [reply]
Thanks for the tips I will use it instead of forcing it through the header.

**Reply by kamcma** [reply]
Glad to find this tip, as I was looking to solve the same thing. I cooked up a small helper to prevent sprinkling `.baseRequestData()`s across router code:
```swift
extension Body where Bytes: ParserPrinter {
  static func fooAPIJSONEncoded<Value: Codable>(_ type: Value.Type) -> BaseURLPrinter<Self>
  where Bytes == Parsers.MapConversion<Parsers.ReplaceError<Rest<Data>>, Conversions.JSON<Value>> {
    self.init(
      .json(
        type,
        decoder: fooAPIConfiguredDecoder,
        encoder: fooAPIConfiguredEncoder
      )
    )
    .baseRequestData(.init(headers: ["Content-Type": ["application/json"]]))
  }
}
```
You can then call like so:

```swift
Route(.case(FooAPI.someRouteWithFooAPICodableTypeBody)) {
  Method.post
  Body.fooAPIJSONEncoded(FooAPICodableType.self)
}
```
Open to suggestions to improve it further; I was mostly just following compiler complaints on the constraints. 🙂

**Reply by kamcma** [reply]
An alternative I pondered but am perhaps not slick enough to implement myself is some sort of higher-order router that would add the `.baseRequestData` to any case matching, eg, POST method and a JSON conversion body. Or something like that. I don't know if that's possible.
